internal/locate/region_cache.go

```diff
@@ -1962,29 +2134,93 @@ func (c *RegionCache) scanRegions(bo *retry.Backoffer, startKey, endKey []byte,
 				util.HexRegionKeyStr(c.codec.EncodeRegionKey(startKey)), util.HexRegionKeyStr(c.codec.EncodeRegionKey(endKey)),
 			)
 		}
-		regions := make([]*Region, 0, len(regionsInfo))
-		for _, r := range regionsInfo {
-			// Leader id = 0 indicates no leader.
-			if r.Leader == nil || r.Leader.GetId() == 0 {
-				continue
-			}
-			region, err := newRegion(bo, c, r)
+		return c.handleRegionInfos(bo, regionsInfo, true)
+	}
+}
+
+// batchScanRegions scans at most `limit` regions from PD, starts from the region containing `startKey` and in key order.
+func (c *RegionCache) batchScanRegions(bo *retry.Backoffer, keyRanges []pd.KeyRange, limit int, opts ...BatchLocateKeyRangesOpt) ([]*Region, error) {
+	if limit == 0 || len(keyRanges) == 0 {
+		return nil, nil
+	}
+	ctx := bo.GetCtx()
+	if span := opentracing.SpanFromContext(ctx); span != nil && span.Tracer() != nil {
+		span1 := span.Tracer().StartSpan("batchScanRegions", opentracing.ChildOf(span.Context()))
+		defer span1.Finish()
+		ctx = opentracing.ContextWithSpan(ctx, span1)
+	}
+	var opt BatchLocateKeyRangesOption
+	for _, op := range opts {
+		op(&opt)
+	}
+	// key info for debug log
+	startKey, endKey := keyRanges[0].StartKey, keyRanges[len(keyRanges)-1].EndKey
+	var backoffErr error
+	for {
+		if backoffErr != nil {
+			err := bo.Backoff(retry.BoPDRPC, backoffErr)
 			if err != nil {
-				return nil, err
+				return nil, errors.WithStack(err)
 			}
-			regions = append(regions, region)
 		}
-		if len(regions) == 0 {
-			return nil, errors.New("receive Regions with no peer")
+		start := time.Now()
+		pdOpts := []pd.GetRegionOption{pd.WithAllowFollowerHandle()}
+		if opt.needBuckets {
+			pdOpts = append(pdOpts, pd.WithBuckets())
+		}
+		regionsInfo, err := c.pdClient.BatchScanRegions(ctx, keyRanges, limit, pdOpts...)
+		metrics.LoadRegionCacheHistogramWithRegions.Observe(time.Since(start).Seconds())
+		if err != nil {
+			if apicodec.IsDecodeError(err) {
+				return nil, errors.Errorf("failed to decode region range key, startKey: %q, limit: %d, err: %v, encode_start_key: %q",
+					util.HexRegionKeyStr(startKey), limit, err, util.HexRegionKeyStr(c.codec.EncodeRegionKey(startKey)))
```

Comment:
Should the `Key` be red-act in error messages?

Reply:
Seems red-act is not supported in current client-go, maybe remove this log and add a TODO by now.